text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+
+### Storing post state hashes
+
+Post state hashes refer to global state, which already forms a tree structure. If the blocks were to store a post state hash inside each block, a client could download only the chain until genesis as additional verification, but then know that the post state hash is correct.
+
+Fetching the associated global state again just becomes a recursive (ideally CDN-powered, at least pack supported) download of the remaining hashes. This willl result in automatic "pruning", since outdated/overwritten data is not retrieved.
+
+This is analogous to a "shallow clone" of a Git repository.
+
+As a whole, this topic is fairly complex and definitely warrants its own RFC.
+
+### Garbage collection
+
+Over time, the node will accumulate some dead weight, especially if the DAG is also stored inside the object store. A garbage collector can be run periodically to remove these keys from the object store backend. This process is also similar to Git, which considers any object pointed by a ref (like `HEAD`, branches or tags) worth keeping and eliminates those objects that are unreachable from any of these *tips*.
+
+When designing this feature, there may be a chance to build in a "shallow" node mode where only the latest global state and minimal history is kept to reduce disk space.
+
+For performance reasons, garbage collection may be run online, that is during normal operation. It should be feasible to keep a cache of all reachable object hashes in memory and keep it up-to-date seperately, especially given our serialized access due to the reactor architecture. With this in place, once the list is completey and guaranteed to be continuously kept up-to-date, the garbage collector can start walking all keys and eliminate unreachable ones.
+
+### In-memory deduplication
+
+Storage could keep a cache of recently used objects and use `Rc`s or `Arcs` to pass these around. This is usefuly because it is likely that a recently received deploy will make its way into a block in the near future. Having an unified mechanism that works for any kind of object makes implementing this case a good investment, as it will immediately benefit the whole system.
+
+### Persisting networking state
+
+Should the networking component to be reworked to gossip data for node discovery, this data will need to be stored as well. It is conceivable that this might be stored to speed up a rejoin-process. This does, however, require a bit of indexing on the storage component's part, as there is currently no way to find all items of a specific object type.
```

Comment:
```suggestion
If the networking component is reworked to gossip data for node discovery, this data will need to be stored as well. It is conceivable that this might be stored to speed up a rejoin-process. This does, however, require a bit of indexing on the storage component's part, as there is currently no way to find all items of a specific object type.
```